diff --git a/diskann-inmem/integration/index/tests.rs b/diskann-inmem/integration/index/tests.rs
index 46fd52f..3aa52b0 100644
--- a/diskann-inmem/integration/index/tests.rs
+++ b/diskann-inmem/integration/index/tests.rs
@@ -44,14 +44,20 @@ pub(super) fn knn(
     let before = index.counters();
     let mut misc = KnnSearch::new();
     let mut neighbors = Vec::new();
+    let mut per_query = Vec::new();
     for (out, query) in std::iter::zip(ids.row_iter_mut(), queries.iter()) {
         neighbors.clear();
 
         let stats = rt.block_on(index.search(query, knn, &mut neighbors))?;
-        misc += stats;
+        misc += stats.clone();
 
         std::iter::zip(out.iter_mut(), neighbors.iter()).for_each(|(d, s)| *d = *s.id());
+        per_query.push(QueryStats {
+            ids: neighbors.iter().map(|n| *n.id()).collect(),
+            misc: stats,
+        });
     }
     let counters = before.delta(&index.counters())?;
@@ -70,6 +76,7 @@ pub(super) fn knn(
         counters,
         recall: recall.into(),
         misc,
+        per_query,
     })
 }
 
@@ -110,11 +117,17 @@ impl std::fmt::Display for KnnRecall {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+pub(crate) struct QueryStats {
+    ids: Vec<u64>,
+    misc: KnnSearch,
+}
+
 #[derive(Debug, Serialize, Deserialize)]
 pub(crate) struct KnnStats {
     recall: KnnRecall,
     counters: Counters,
     misc: KnnSearch,
+    per_query: Vec<QueryStats>,
 }
